src/common/telemetry.ts

```diff
@@ -0,0 +1,293 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Gitpod. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as os from 'os';
+import { Analytics, AnalyticsSettings } from '@segment/analytics-node';
+import { ILogService } from '../services/logService';
+import { cloneAndChange, escapeRegExpCharacters, mixin } from '../common/utils';
+
+export const ProductionUntrustedSegmentKey = 'untrusted-dummy-key';
+
+
+export const TRUSTED_VALUES = new Set([
+	'gitpodHost'
+]);
+
+export interface TelemetryEventProperties {
+	gitpodHost: string;
+
+	[key: string]: any;
+}
+
+export interface UserFlowTelemetryProperties {
+	flow: string;
+
+	gitpodHost: string;
+	gitpodVersion?: string;
+
+	workspaceId?: string;
+	instanceId?: string;
+
+	userId?: string;
+
+	[key: string]: any;
+}
+
+export interface ITelemetryService {
+	sendTelemetryEvent(eventName: string, properties?: TelemetryEventProperties): void;
+	sendTelemetryException(error: Error, properties?: TelemetryEventProperties): void;
+
+	sendUserFlowStatus(status: string, flow: UserFlowTelemetryProperties): void;
+}
+
+export function createSegmentAnalyticsClient(logService: ILogService, gitpodHost: string, segmentKey: string): Analytics | undefined {
+    const serviceUrl = new URL(gitpodHost);
+
+    const settings: AnalyticsSettings = {
+        writeKey: segmentKey,
+        // in dev mode we report directly to IDE playground source
+        host: 'https://api.segment.io',
+        path: '/v1/batch'
+    };
```

Comment:
By default segment buffer the events and then send them when the buffer is full/some time pass, this is fine in the extension because we can flush in the `deactivate` method, for proxy.ts we should disable the buffer and send the events immediately as is the ssh client that manages the lifecycle